scapy/contrib/ife.py

```diff
@@ -0,0 +1,110 @@
+
+"""
+    IFE - ForCES Inter-FE LFB type
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    :author:    Alexander Aring, aring@mojatatu.com
+    :license:   GPLv2
+
+        This module is free software; you can redistribute it and/or
+        modify it under the terms of the GNU General Public License
+        as published by the Free Software Foundation; either version 2
+        of the License, or (at your option) any later version.
+
+        This module is distributed in the hope that it will be useful,
+        but WITHOUT ANY WARRANTY; without even the implied warranty of
+        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+        GNU General Public License for more details.
+
+    :description:
+
+        This module provides Scapy layers for the IFE protocol.
+
+        normative references:
+            - RFC 8013
+              Forwarding and Control Element Separation (ForCES)
+              Inter-FE Logical Functional Block (LFB)
+              https://tools.ietf.org/html/rfc8013
+"""
+
+from scapy.data import ETHER_TYPES
+from scapy.packet import Packet, bind_layers
+from scapy.fields import FieldLenField, PacketListField, IntField, \
+    ShortField, PadField, StrNullField
+from scapy.layers.l2 import Ether
+
+ETH_P_IFE = 0xed3e
+ETHER_TYPES['IFE'] = ETH_P_IFE
+
+
+class IFETlv(Packet):
+    """
+    Parent Class interhit by all ForCES TLV strucutures
+    """
+    pass
+
+
+class IFE(Packet):
+    """
+    Main IFE Packet Class
+    """
+    name = "IFE"
+
+    META_SKBMARK = 1
+    META_HASHID = 2
+    META_PRIO = 3
+    META_QMAP = 4
+    META_TCINDEX = 5
+
+    fields_desc = [FieldLenField("mdlen", None, length_of="tlvs",
+                                 adjust=lambda pkt, x: x+2),
+                   PacketListField("tlvs", None, IFETlv),
+                  ]
+
+
+def gen_ife_tlv_fields(ife_type, value_cls, value):
+    """
+    Runtime generation of fields_desc for IFE TLV classes
+    """
+    return [ShortField("type", ife_type),
+            FieldLenField("length", None, length_of="value",
+                          adjust=lambda pkt, x: x+4),
+            PadField(value_cls("value", value), 4, padwith=b'\x00'),
+           ]
+
+
+class IFESKBMark(IFETlv):
+    """
+    Linux metadata field for skb mark field
+    """
+    fields_desc = gen_ife_tlv_fields(IFE.META_SKBMARK, IntField, 0)
+
+
+class IFEPrio(IFETlv):
+    """
+    Linux metadata field for TC prio
+    """
+    fields_desc = gen_ife_tlv_fields(IFE.META_PRIO, IntField, 0)
+
+
+class IFETCIndex(IFETlv):
+    """
+    Linux metadata field for TC index
+    """
+    fields_desc = gen_ife_tlv_fields(IFE.META_TCINDEX, ShortField, 0)
+
+
+class IFEString(IFETlv):
+    """
+    Just a TLV with variable input size and point to IFE.META_SKBMARK
+    to let the kernel accept it
+    """
+    # Let it be IFE.META_SKBMARK for now so the Kernel will parse it
+    fields_desc = gen_ife_tlv_fields(IFE.META_SKBMARK, StrNullField, "")
```

Comment:
If all those sub-layers are that similar, please consider merging them in one more generic layer.
If you do hat and apply the other comments, it will look much cleaner